### ISSUE TITLE
Python vs. BrainScript: same network, but different training behavior

### ISSUE THREAD

**Issue by GhMaRo** [Microsoft/CNTK#3011]
I am trying to build a Python network which should be identical to my existing BrainScript network. In principle, I'm already done with it, but the behavior is not really the same. The BrainScript network trains much faster and I don't know why. The training loss of both networks is nearly the same at the beginning but then the loss of the BrainScript network decreases much faster than the loss of the Python network. 

In BrainScript, I only create the initial model and then I load this model in Python and train it there. So both networks are using the same Python code while training (same learningRate, same input data,...). Only the initial models can be different but should be the same. I have already plotted the graphs of both models and checked them: In principle I did not find any difference, but the layers/functions are plottet differently. E.g. a BatchNormalizationLayer does not look the same in Python as in BrainScript.

Do Python and BrainScript use the same C++ functions internally?

These are the most important functions/layers which I am using:

```
# Python
Convolution((k,k), num_filters = outMap, init = he_normal(scale = 0.141421), pad = True, strides = (s,s), bias = True)
Convolution((k,k), num_filters = outMap, init = he_normal(scale = 0.141421), pad = True, dilation = (d,d), bias = True)
BatchNormalization(map_rank = 1, init_scale = 1, normalization_time_constant = 0, epsilon = 0.00001, use_cntk_engine = False)
relu
Dropout(dropoutRate)
convolution(kernel, input)
		
# BrainScript	
ConvolutionalLayer(outMap, (k:k), pad=true, stride=(s:s), init="heNormal", initValueScale = 0.141421,  bias=true)
ConvolutionalLayer(outMap, (k:k), pad=true, init="heNormal", initValueScale = 0.141421, bias=true, dilation = (d:d:1:1))
BatchNormalizationLayer(spatialRank = 2, normalizationTimeConstant = 0, initialScale = 1, epsilon = 0.00001, useCntkEngine = false)
RectifiedLinear
Dropout
Convolution(kernel, input, kernelSize:kernelSize, mapDims = 1)
```
In my opinion, the listed functions/layers should be identical. Or did I miss an important parameter which I have to set (or sth. else)?

The only difference should be the random seed, but I can't believe that this make such a big difference in training behaviour.

I am greatful for any hints, why the behavior could be different.

**Comment by KeDengMS** [comment]
Your Python script uses layers API which wraps Convolution in a block function. Block functions are shown in plot as a single node, while BrainScript does not do so. Please check the input/output dimensions in the plot, as well as parameter sizes/counts between these two, which should match.

**Comment by GhMaRo** [comment]
I have already checked input/output dimensions as well as the parameter sizes/counts.

With my question "**Do Python and BrainScript use the same C++ functions internally?**" I actually wanted to know if the layers and functions of BrainScript and Python (e.g. convolution, BatchNormalization etc.) use the same C++ code? Or are they implemented in a different way?

I ask this, because a few days ago [this issue](https://github.com/Microsoft/CNTK/issues/2990) was opened. So in case that for example the BatchNormalization is not implemented in the same way for Python and BrainScript, this could maybe explain the different behavior.

**Comment by KeDengMS** [comment]
They are implemented with the same C++ function if running on the same cpu/gpu device. The ONNX batch norm issue seems to be a different one and I'll have some colleagues working on ONNX to take a look.

**Comment by GhMaRo** [comment]
Hm ok, thanks. Then there must something else be different between the networks.

Another question to the plotting of the graphs:

I load the created initial models (Python / BrainScript) in both cases with my Python script and plot them both with C.logging.graph.plot(). Is there a way to "convert" the loaded BrainScript model to a Python model so that it will be plottet in the same way as the **in Python created model**?

Also when I train the loaded BrainScript model with my Python script and then save this trained model, it remains a "BrainScript model" (according to the plottet graph). So internally there must something be different between the **in BrainScript created model** and the **in Python created model**.

**Comment by KeDengMS** [comment]
If you don't use layers API in Python, there will be no block functions. Or you can clone the graph and replace the blocks with low level APIs like [convolution](https://cntk.ai/pythondocs/cntk.ops.html#cntk.ops.convolution).

If you have suspicion about BrainScript/Python difference, one way to find out is to store a BrainScript model after training for a while, and then load the model in both BS and Python to continue the training using the same dataset. I'd expect the results to match between these two, as the they share the same C++ implementation.
